[AI generation] Identify ISO language code and track it

selanthiraiyan opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #10103 
<!-- Id number of the GitHub issue this PR addresses. -->

## Description

Identifies the ISO language code instead of name and tracks that upon successful AI generation.

- The prompt is updated to identify the language code instead of the name.
- The language code is tracked when the product description or product sharing message is generated.

## Testing instructions

**Product Description**
- Log in to a published WPCom site.
- Select an existing product or create a new one.
- Select Write with AI in the description row.
- On the description AI sheet, select Write with AI. The response should be in the same language as the product title and features.
- Check Xcode logs and ensure that the following event is tracked with the `language` and `source` properties
```
Tracked ai_identify_language_success, properties: [AnyHashable("is_wpcom_store"): true, AnyHashable("language"): "en", AnyHashable("blog_id"): 12345, AnyHashable("source"): "product_description"]
```
- Tap Regenerate and ensure that the generated description is still good.
- Dismiss the sheet. Turn off the internet.
- Select Write with AI in the description row.
- On the description AI sheet, select Write with AI.
- Check Xcode logs and ensure that the following event is tracked with the `language` and `source` properties
```
Tracked ai_identify_language_failed, properties: [AnyHashable("error_description"): "<long error description here>", AnyHashable("source"): "product_description", AnyHashable("error_domain"): "NSURLErrorDomain", AnyHashable("error_code"): "-1009", AnyHashable("blog_id"): 12345, AnyHashable("is_wpcom_store"): true]
```

**Product Sharing message**
- Log in to a published WPCom site.
- Select an existing product or create a new one.
- Tap the `...` button from the top right and tap Share to open the share sheet
- Tap "Write with AI" to generate a "share message".
- Check Xcode logs and ensure that the following event is tracked with the language code as the value for `language` property
```
Tracked ai_identify_language_success, properties: [AnyHashable("language"): "en", AnyHashable("is_wpcom_store"): true, AnyHashable("source"): "product_sharing", AnyHashable("blog_id"): 12345]
```
- The generated share message should be in the same language as the product title and description.
- Tap Regenerate and ensure that the message is still good.
- Dismiss the sheet. Turn off the internet.
- Tap the `...` button from the top right and tap Share to open the share sheet
- Tap "Write with AI" to generate a "share message".
- Check Xcode logs and ensure that the following event is tracked with the `language` and `source` properties
```
Tracked ai_identify_language_failed, properties: [AnyHashable("error_domain"): "NSURLErrorDomain", AnyHashable("error_code"): "-1009", AnyHashable("error_description"): "<Long error description here>", AnyHashable("blog_id"): 12345, AnyHashable("source"): "product_sharing", AnyHashable("is_wpcom_store"): true]
```

## Screenshots
| Product description | Product sharing |
|--------|--------|
| ![Simulator Screen Shot - iPhone 14 Pro - 2023-07-05 at 10 28 26](https://github.com/woocommerce/woocommerce-ios/assets/524475/92c8129a-a6e0-445d-9cf1-0478778b8169) | ![Simulator Screen Shot - iPhone 14 Pro - 2023-07-05 at 10 28 15](https://github.com/woocommerce/woocommerce-ios/assets/524475/fd538127-befe-42cf-9c88-62fb9fd16da6) |

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

selanthiraiyan committed:
Update language identification prompt to return ISO language code instead of language name.

selanthiraiyan committed:
Track the identified language after successful description generation.

selanthiraiyan committed:
Track the identified language after successful product sharing message generation.

selanthiraiyan committed:
Include all relevant PRs in release notes.

itsmeichigo reviewed (commented):
The new change still works for my test store and products 👍  

Regarding the tracking: I think we should add a separate event for the identified language when we get the response from AI. Including the property in the text generation success event seems misleading as we don't identify the language as part of that request. WDYT?

selanthiraiyan committed:
New events for identify language success and failure.

selanthiraiyan committed:
Track identify language success and failure events during product description generation.

selanthiraiyan committed:
Track identify language success and failure events during product sharing message generation.

selanthiraiyan committed:
Merge branch 'trunk' into bug/10103-language-to-tracks

selanthiraiyan committed:
Fix typo.

selanthiraiyan commented:
> Regarding the tracking: I think we should add a separate event for the identified language when we get the response from AI. Including the property in the text generation success event seems misleading as we don't identify the language as part of that request. WDYT?

Good point. I agree, @itsmeichigo . 👍 

- I have introduced new tracking events for identifying language events in https://github.com/woocommerce/woocommerce-ios/pull/10179/commits/6626a1fe08714a47b9572537bbc4f2bd386e0518.
- Made the necessary changes to track and unit test the new event tracking. 
- Updated the testing steps in the PR description to reflect the new events. 

Kindly take another look at the PR.

itsmeichigo commented:
Sorry for the many comments @selanthiraiyan - but I'm thinking of unifying the tracking for language detection as this uses the same API request for product description and sharing messages. That way we can avoid creating more Tracks events than necessary.

How about using one single event `*_ai_languge_detection` with properties: 
- `is_success`: true | false
- `source`: `product_description` | `product_sharing`
- `error`: optional error information.

selanthiraiyan commented:
Thanks, @itsmeichigo!

Please keep the comments coming. Getting the tracking right is important; your perspective is super helpful in achieving that. 🙏  

I added separate events because we use separate tracking events for text generation. I agree that we can use a single event for identify language as we use the same `ProductAction.identifyLanguage` action in both screens.   

---

Some thoughts about the proposed event.

1. As we have been using the word "identify" in the code, I think we should stick to using "identify" instead of using "detection". 
2. We have to add the identified language code to the event. I will use the key as `language` for that property.
3. I have seen that we usually use two separate events for success and failure cases. 
   1. Considering that the `language` property will be only available for success scenarios, do you think it is worth it to separate these into two events?
   2. Will having a single event with a property which can also be `nil` limit us when we try to measure this using Funnels?

**A. Two events**

Event name - `woocommerceios_ai_identify_language_success`

- `source`: `product_description` | `product_sharing`
- `language`: Identified language code

Event name - `woocommerceios_ai_identify_language_failure`

- `source`: `product_description` | `product_sharing`
- `error`: error information.

**B. Single event**

Event name - `woocommerceios_ai_identify_language`

- `is_success`: true | false
- `source`: `product_description` | `product_sharing`
- `error`: optional error information.
- `language`: optional language code information

Kindly let me know your thoughts.

itsmeichigo commented:
I agree that the 2 event approach looks clearer 👍  Please go ahead with that solution, thanks!

selanthiraiyan committed:
New event for tracking identify language success and failure cases.

selanthiraiyan committed:
Use same identify language events for product description and sharing message flows.

selanthiraiyan committed:
Remove unused analytics events.

selanthiraiyan committed:
Update unit tests to test for new identify language event.

selanthiraiyan committed:
Update comments.

selanthiraiyan committed:
Update comments to use identify instead of detection.

selanthiraiyan commented:
@itsmeichigo 
- I made the necessary changes to use the same tracking event on both product description and product sharing views.
- I have also updated the testing instructions. 
- After we merge this PR, I will update the respective PTs and inform the Android team about it. 

Kindly take another look at the PR. Thank you!

itsmeichigo reviewed (approved):
Thanks for the updates, code looks good and tests well :shipit:
